packages/kbn-reporting/public/job_completion_notifications.ts

```diff
@@ -9,29 +9,55 @@
 import { JOB_COMPLETION_NOTIFICATIONS_SESSION_KEY } from '@kbn/reporting-common';
 import { JobId } from '@kbn/reporting-common/types';
 
-const set = (jobs: string[]) => {
-  sessionStorage.setItem(JOB_COMPLETION_NOTIFICATIONS_SESSION_KEY, JSON.stringify(jobs));
-};
+// Reading and writing to the local storage must be atomic,
+// i.e. performed in a single operation. This storage queue
+// Operations on the localStorage key can happen from various
+// parts of code. Using a queue to manage async operations allows
+// operations to process one at a time
+let operationQueue = Promise.resolve();
+async function addToQueue(func: (error: Error | null) => void) {
+  operationQueue = operationQueue.then(() => func(null)).catch(func);
+  await operationQueue;
+}
 
-const getAll = (): string[] => {
-  const sessionValue = sessionStorage.getItem(JOB_COMPLETION_NOTIFICATIONS_SESSION_KEY);
-  return sessionValue ? JSON.parse(sessionValue) : [];
-};
+export async function getPendingJobIds(): Promise<JobId[]> {
+  let jobs: JobId[] = [];
+  await addToQueue(async () => {
+    // get the current jobs
+    const jobsData = localStorage.getItem(JOB_COMPLETION_NOTIFICATIONS_SESSION_KEY);
+    jobs = jobsData ? JSON.parse(jobsData) : [];
+  });
+  return jobs;
+}
 
-export const add = (jobId: JobId) => {
-  const jobs = getAll();
-  jobs.push(jobId);
-  set(jobs);
-};
+export async function addPendingJobId(jobId: JobId) {
+  await addToQueue(async (error: Error | null) => {
+    return new Promise((resolve, reject) => {
+      if (error) {
+        window.console.error(error);
+        reject(error);
+      }
+      // get the current jobs synchronously
+      const jobsData = localStorage.getItem(JOB_COMPLETION_NOTIFICATIONS_SESSION_KEY);
+      const jobs: JobId[] = jobsData ? JSON.parse(jobsData) : [];
+      // add the new job
+      jobs.push(jobId);
+      // write back to local storage
+      localStorage.setItem(JOB_COMPLETION_NOTIFICATIONS_SESSION_KEY, JSON.stringify(jobs));
```

Comment:
what is instead of managing the state as a single blob, we would write a job state for each job separately. 
e.g. 

localStorage.setItem(`prefix-${jobId}`, job), 
would this keep things simpler?

Reply:
It would complicate the process of retrieving all of the pending jobs. We can use a common prefix, with a dynamic suffix for the job ID. But to get all of the pending job IDs, we need to get all of the keys from local storage and check each one to see if it has the matching prefix. So, I think this would not make the process simpler.

Reply:
> But to get all of the pending job IDs, we need to get all of the keys from local storage and check each one to see if it has the matching prefix.

yes, but this keeps the code synchronous, and there is no need to manage the queue. it seems like this option would be simpler even with the need to read all keys (which would be abstracted in a a small function)

Reply:
Since there won't be millions of keys in local storage, this seems fair. I think this is good feedback since the change to async wasn't needed for the important part of this fix.

Pushed a471462cb72